File: apps/write-8-minutes/ios/Anky/Features/Painting/LevelTheme.swift
import SwiftUI

/// Palette-driven theming: each level's painting tints the whole app.
///
/// Derived from the package's 4-6 swatches (dark → light): a background wash,
/// a glow tint for the frame, and a button warmth. Falls back to the lazure
/// sepia register when no painting is installed. Never pure black.
struct LevelTheme: Equatable {
    let swatches: [Color]
    let backgroundWash: Color
    let glowTint: Color
    let buttonWarmth: Color

    static let fallback = LevelTheme(
        swatches: [],
        backgroundWash: .ankyPaperDeep,
        glowTint: .ankyGold,
        buttonWarmth: .ankyGoldLight
    )

    init(swatches: [Color], backgroundWash: Color, glowTint: Color, buttonWarmth: Color) {
        self.swatches = swatches
        self.backgroundWash = backgroundWash
        self.glowTint = glowTint
        self.buttonWarmth = buttonWarmth
    }

    init(palette: [String]) {
        let colors = palette.compactMap(Color.fromHexSwatch)
        guard colors.count >= 3 else {
            self = .fallback
            return
        }
        // Palette arrives sorted dark → light.
        swatches = colors
        // The wash leans on the midtone, lifted toward parchment so text
        // stays readable and the darkness stays warm.
        let mid = colors[colors.count / 2]
        backgroundWash = mid.opacity(0.35)
        // Glow comes from the brightest swatch — usually the painting's gold.
        glowTint = colors[colors.count - 1]
        buttonWarmth = colors[max(0, colors.count - 2)]
    }

    init(package: PaintingPackage?) {
        if let package {
            self.init(palette: package.palette)
        } else {
            self = .fallback
        }
    }

    /// The LazureWall mood carrying this level's pigment.
    var wallMood: LazureWall.Mood {
        swatches.isEmpty ? .dawn : .kingdom(glowTint)
    }
}

extension Color {
    /// Parses "#rrggbb" into a Display P3 color, matching the lazure system.
    static func fromHexSwatch(_ hex: String) -> Color? {
        var value = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        guard value.hasPrefix("#") else { return nil }
        value.removeFirst()
        guard value.count == 6, let raw = UInt32(value, radix: 16) else { return nil }
        return Color(
            .displayP3,
            red: Double((raw >> 16) & 0xFF) / 255,
            green: Double((raw >> 8) & 0xFF) / 255,
            blue: Double(raw & 0xFF) / 255
        )
    }
}

File: apps/android/app/src/main/java/inc/anky/android/ui/lazure/LevelTheme.kt
package inc.anky.android.ui.lazure

import androidx.compose.runtime.Immutable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.colorspace.ColorSpaces

/**
 * Palette-driven theming: each level's painting tints the whole app.
 *
 * Port of the iOS `LevelTheme`. Derived from a painting package's 4-6
 * swatches (sorted dark -> light): a background wash, a glow tint for the
 * frame, and a button warmth. Falls back to the lazure sepia register when
 * no painting is installed. Never pure black.
 *
 * Pure data + pure derivation: no painting-package type is imported here.
 * Callers hand over the package's `palette` hex strings via [fromPalette].
 */
@Immutable
data class LevelTheme(
    val swatches: List<Color>,
    val backgroundWash: Color,
    val glowTint: Color,
    val buttonWarmth: Color,
) {

    /** The [LazureWall] mood carrying this level's pigment. */
    val wallMood: LazureMood
        get() = if (swatches.isEmpty()) LazureMood.Dawn else LazureMood.Kingdom(glowTint)

    companion object {
        /** iOS `LevelTheme.fallback` — the lazure sepia register. */
        val Fallback = LevelTheme(
            swatches = emptyList(),
            backgroundWash = LazurePigments.ankyPaperDeep,
            glowTint = LazurePigments.ankyGold,
            buttonWarmth = LazurePigments.ankyGoldLight,
        )

        /**
         * Derives the theme from a painting's meta palette ("#rrggbb"
         * strings, sorted dark -> light). Mirrors the iOS init exactly:
         *  - fewer than 3 parseable swatches -> [Fallback];
         *  - the wash leans on the midtone at 0.35 alpha, lifted toward
         *    parchment so text stays readable and the darkness stays warm;
         *  - glow comes from the brightest swatch — usually the painting's
         *    gold;
         *  - button warmth is the second-brightest.
         */
        fun fromPalette(palette: List<String>): LevelTheme {
            val colors = palette.mapNotNull(::colorFromHexSwatch)
            if (colors.size < 3) return Fallback
            val mid = colors[colors.size / 2]
            return LevelTheme(
                swatches = colors,
                backgroundWash = mid.copy(alpha = 0.35f),
                glowTint = colors[colors.size - 1],
                buttonWarmth = colors[maxOf(0, colors.size - 2)],
            )
        }
    }
}

/**
 * Parses "#rrggbb" into a Display P3 color, matching the lazure system
 * (port of the iOS `Color.fromHexSwatch`). Returns null unless the string
 * is exactly a leading `#` and six hex digits (surrounding whitespace
 * tolerated, like the Swift version).
 */
fun colorFromHexSwatch(hex: String): Color? {
    val value = hex.trim()
    if (!value.startsWith("#")) return null
    val body = value.drop(1)
    if (body.length != 6) return null
    val raw = body.toLongOrNull(radix = 16) ?: return null
    return Color(
        red = ((raw shr 16) and 0xFF).toFloat() / 255f,
        green = ((raw shr 8) and 0xFF).toFloat() / 255f,
        blue = (raw and 0xFF).toFloat() / 255f,
        alpha = 1f,
        colorSpace = ColorSpaces.DisplayP3,
    )
}
